### ISSUE TITLE
Correctly mark all symbols as `public` or not

### ISSUE THREAD

**Issue by LilithHafner** [JuliaLang/julia#51335]
#50105 added a preliminary list of symbols to be marked as `public` from `Base`. We should finalize that list for Base, stdlibs, and submodules before 1.11.

**Comment by Brakkugit** [comment]
is this issue solved or can i help?

**Comment by LilithHafner** [comment]
This issue is very much not solved, and we would appreciate your help!

Any symbol that is mentioned in the manual in Julia 1.9 or 1.10 and is not exported should be marked as public with `public that_symbol`. If you want to help, you can look through submodules of Base and/or standard libraries to find these symbols and mark them as public. Ideally, you would find many symbols and mark them all as public in a single PR so discussion can be more effective. In some cases it might not be clear whether or not a symbol is supposed to be public, feel free to take a guess and mention that you are unsure and other folks can weigh in with their opinions.

**Comment by Brakkugit** [comment]
Can you explain me bit more specifically about how should i approach this.

**Comment by Brakkugit** [comment]
so should i download all the dependencies for building julia or can i just clone it in my local machine and run for this task

**Comment by LilithHafner** [comment]
Ideally, this should work:
```
$ git clone https://github.com/JuliaLang/julia
$ cd julia
$ make
... wait a long time ...
$ ./julia
```
If you try this and get an error and you are running on a mainstream OS, let us know, because that's probably a bug on our end.

**Comment by Brakkugit** [comment]
make doesnt work for windows

**Comment by LilithHafner** [comment]
Ah, right. Windows. The steps for building from source are a bit more complicated: https://docs.julialang.org/en/v1/devdocs/build/windows/#Source-distribution

You don't have to build from source to contribute, but it can be very helpful in testing your PRs.

**Comment by nsajko** [comment]
`Base.include` is documented in the manual and referred to from other help sections, so it seems like it's part of the public interface. The REPL help, however, says: "The following bindings may be internal; [...]"

**Comment by vtjnash** [comment]
Every module has an `include` function that is private to that module

**Comment by nsajko** [comment]
`Base.include` and `Base.MainInclude.include` are two different functions, both documented. The *latter* is the one exported from `Base`:

```julia-repl
julia> Base.MainInclude.include === include
true
```

**Comment by inkydragon** [comment]
What about Stdlibs?

Take `SHA.jl` as an example.
All symbols list in docs are exported, so there is nothing to do, right?

docs HTML: https://docs.julialang.org/en/v1/stdlib/SHA/
docs md: https://github.com/JuliaCrypto/SHA.jl/blob/88e1c83c32790c1cfbd0d13cc71c7db32bd6a573/docs/src/index.md?plain=1#L61
export: https://github.com/JuliaCrypto/SHA.jl/blob/88e1c83c32790c1cfbd0d13cc71c7db32bd6a573/src/SHA.jl#L36-L49

Do we need to change `export` to `public` or something else?

**Comment by nsajko** [comment]
> Do we need to change `export` to `public` or something else?

No need to do that, AFAIK: either one of `public` or `export` denote public API, the difference is just that public doesn't pollute namespaces (when imported with `using` instead of `import`).

**Comment by ShrutiRDalvi** [comment]
Hello, 

Is this issue solved, else I'd love to contribute. It is my first time contributing to open source, so any tips on how to go about it would help a ton. 

Thanks

**Comment by lgoettgens** [comment]
For the `Tar` stdlib, @stevengj made a proposal in https://github.com/JuliaIO/Tar.jl/pull/173#discussion_r1480230833

**Comment by mcabbott** [comment]
> Submodules of Base

I note that Base.Iterators exports everything:

https://github.com/JuliaLang/julia/blob/master/base/iterators.jl#L38

yet they still show as private:
```
help?> Iterators.map
  │ Warning
  │
  │  The following bindings may be internal; they may change or be removed in future
  │  versions:
  │
  │    •  Base.Iterators.map

  Iterators.map(f, iterators...)

  Create a lazy mapping. This is another syntax for writing (f(args...) for args in
  zip(iterators...)).
```

Adding explicit `public` within the module does not change this. And the docs for `public` make it sound like this is very much a per-module concept:

https://docs.julialang.org/en/v1.11-dev/manual/modules/#modules

If in fact it's a package-level concept, then I'd expect either `public Iterators.map` and `public .Iterators.map` to work. But these are both syntax errors.

**Comment by matthias314** [comment]
I'm looking at candidates in `Base.Broadcast`. Where should the corresponding `public` statement go? To broadcast.jl (following `export`?), to public.jl or elsewhere?

**Comment by matthias314** [comment]
OK, the PR for broadcast.jl is #54060.

**Comment by KristofferC** [comment]
I don't think this has to be on the milestone since adding a `public` can always be done in a later release.

**Comment by LilithHafner** [comment]
It's just that users will start getting "this symbol may be internal" warnings for docstrings of nonpublic symbols in 1.11

**Comment by nsajko** [comment]
`Base.dataids` should be `public`, right? It's doc string says:

> Custom arrays that would like to opt-in to aliasing detection of their component parts can specialize this method to return the concatenation of the `dataids` of their component parts. A typical definition for an array that wraps a parent is `Base.dataids(C::CustomArray) = dataids(C.parent)`.

**Comment by CameronBieganek** [comment]
> I don't think this has to be on the milestone since adding a `public` can always be done in a later release.

I'm not a fan of this, because that means a symbol could be public on 1.10 (because it's in the manual), private on 1.11 (because it's not declared with `public`), and then public again on 1.12 (because it gets declared with `public` later on). We should be encouraging less usage of private symbols, not more. Anything on 1.11 that is not declared with `export` or `public` is [*officially* private](https://github.com/JuliaLang/julia/blob/v1.11.0-beta1/doc/src/manual/faq.md#public-apiid-man-api).

In fact, making a public symbol private in a minor release is a breaking change according to SemVer, so this issue should block release of 1.11.

**Comment by CameronBieganek** [comment]
There are some modules that are declared `public`:

```julia
public
# Modules
    Checked,
    Filesystem,
    Order,
    Sort,
```

Should the `Iterators` and `ScopedValues` modules be added to that list? (Is ScopedValues in the 1.11 release?)

**Comment by LilithHafner** [comment]
> `Iterators`

Iterators is already public by way of export

```julia
julia> Iterators
Base.Iterators

julia> Base.ispublic(Base, :Iterators)
true

julia> VERSION
v"1.11.0-beta1"
```

> `ScopedValues`

Either that module should be public or Base should re-publicize its exports. I prefer the former, slightly.

**Comment by CameronBieganek** [comment]
> Iterators is already public by way of export

Ah, I see, thanks.

> > `ScopedValues`
>
> Either that module should be public or Base should re-publicize its exports. I prefer the former, slightly.

Yeah, I think the `ScopedValues` module should be made "public" (as opposed to `public`) by being added to the `export` list for `Base` (along with `Iterators`, `Threads`, `Sys`, etc).
